spacy_llm/backends/rest/backend/base.py

```diff
@@ -69,31 +74,46 @@ def credentials(self) -> Dict[str, str]:
         """
 
     def retry(
-        self,
-        call_api: Callable[[], requests.Response],
+        self, call_method: Callable[..., requests.Response], url: str, **kwargs
     ) -> requests.Response:
         """Retry a call to an API if we get a non-ok status code.
         This function automatically retries a request if it catches a response
-        with an error code in `error_codes`. The amount of timeout also increases
+        with an error code in `error_codes`. The time interval also increases
         exponentially every time we retry.
-        call_api (Callable[[], requests.Response]): Call to API.
-        error_codes (Tuple[int]): Error codes indicating unsuccessful calls.
+        call_method (Callable[[str, ...], requests.Response]): Method to use to fetch request. Must accept URL as first
+            parameter.
+        url (str): URL to address in request.
+        kwargs: Keyword args to be passed on to request.
         RETURNS (requests.Response): Response of last call.
         """
-        timeout = self._timeout
-        response = call_api()
+
+        def _call_api() -> requests.Response:
+            """Calls API with given timeout.
+            RETURNS (requests.Response): Response object.
+            """
+            try:
+                return call_method(url, **kwargs)
+            except (ConnectTimeout, ReadTimeout, TimeoutError) as err:
+                raise TimeoutError(
+                    "Request time out. Check your network connection and the API's availability."
+                ) from err
+
+        interval = self._interval
+        response = _call_api()
         i = 0
 
         # We don't want to retry on every non-ok status code. Some are about
         # incorrect inputs, etc. and we want to terminate on those.
         while i < self._max_tries and response.status_code in self._retry_error_codes:
-            time.sleep(timeout)
+            time.sleep(interval)
+            response = _call_api()
             i += 1
-            timeout = timeout * 2  # Increase timeout everytime you retry
+            # Increase timeout everytime you retry
+            interval = interval * 2
 
         if response.status_code in self._retry_error_codes:
             raise ConnectionError(
-                f"API could not be reached within {self._timeout} seconds in {self._max_tries} attempts. Check your "
+                f"API could not be reached within {self._interval} seconds in {self._max_tries} attempts. Check your "
```

Comment:
I'm wondering if there is a way we can reflect the fact that there was an exponential back-off? Could we document it in the docstrings? (I'll add a suggestion)

Reply:
Rephrased the error message in https://github.com/explosion/spacy-llm/pull/110/commits/0e8a3ca68d22ad96f9e1be5a8873d72c07cd7a9e. Better?

Reply:
Looks great!